File: src/python/solutions/21.py
from typing import List, Optional, Dict, Set, Tuple
import collections
import heapq
import itertools
import math

# Local include
from inc.test import assertTest
import inc.debug as db
from inc.listnode import *

"""
Problem 21: Merge Two Sorted Lists
LeetCode: https://leetcode.com/problems/merge-two-sorted-lists/
:explain: 
"""

INFO = True
APR = 1
if APR == 1:
    apr_idea = "Dynamic Programming"
    time_cmplx = "m+n"
    space_cmplx = "1"
elif APR == 2:
    apr_idea = "Recursion"
    time_cmplx = "m+n"
    space_cmplx = "m+n"

db.DEBUG = True
# Implement
class Solution(object):
    if APR == 1:
        def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
            dummy = ListNode()
            cur = dummy

            while list1 and list2:
                if list1.val > list2.val:
                    cur.next = list2
                    list2 = list2.next
                else:
                    cur.next = list1
                    list1 = list1.next
                
                cur = cur.next
            
            # Merge the rest of longer list
            if list1:
                cur.next = list1
            else:
                cur.next = list2
            
            return dummy.next
    elif APR == 2:
        def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
            if not list1 or not list2:
                return list1 if list1 else list2

            if list1.val > list2.val:
                list1, list2 = list2, list1

            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1

def test21():
    if INFO:
        print(f"\033[35m=========== INFO ===========\033[0m")
        print(f"FILE: {__file__}")
        print(f"APPROACH {APR}: {apr_idea}")
        print(f"  + Time Complexity: O({time_cmplx})")
        print(f"  + Space Complexity: O({space_cmplx})")

    print("\033[35m========== TESTCASE ========\033[0m")
    class Case:
        def __init__(self, list1, list2, exp):
            #Inputs
            self.list1 = list1
            self.list2 = list2
            #Expected
            self.exp = exp

    cases = [
        ([1,2,4], [1,3,4], [1,1,2,3,4,4]),
        ([], [], []),
        ([], [0], [0])
    ]

    sol = Solution()

    for i, (list1, list2, exp) in enumerate(cases):
        list1 = ListNode.create_from_list(list1)
        list2 = ListNode.create_from_list(list2)
        exp = ListNode.create_from_list(exp)
        inputs = [list1, list2]
        res = sol.mergeTwoLists(list1, list2)

        if assertTest(res, exp, i) == False:
            print(f"   Input    : ", end="")
            for input in inputs:
                print(f"{input}", end=" | ")
            print()

if __name__ == "__main__":
    test21()

File: src/cpp/solutions/21.cpp
#include "../inc/global.h"
using namespace std;

// Turn off debug mode (uncomment to turn on)
#undef DEBUG

// Display info of the approach
#define INFO
#define APR 1
#ifdef INFO
#if APR == 1
string apr_idea = "Dynamic Programming";
string time_cmplx = "m+n";
string space_cmplx = "1";
#elif APR == 2
string apr_idea = "Recursion";
string time_cmplx = "m+n";
string space_cmplx = "m+n";
#endif
#endif

/**
 * Problem 21: Merge Two Sorted Lists
 * LeetCode: https://leetcode.com/problems/merge-two-sorted-lists/
 * @input:
 * @output:
*/
class Solution {
public:
    #if APR == 1
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode(0);
        ListNode* cur = dummy;

        while (list1 && list2) {
            if (list1->val > list2->val) {
                cur->next = list2;
                list2 = list2->next;
            } else {
                cur->next = list1;
                list1 = list1->next;
            }
            cur = cur->next;
        }

        cur->next = list1 ? list1 : list2;

        ListNode* head = dummy->next;
        delete dummy;
        return head;
    }
    #elif APR == 2
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (!list1 || !list2) {
            return list1 ? list1 ; list2;
        }

        if (list1->val > list2->val) {
            swap(list1, list2);
        }

        list1->next = mergeTwoLists(list1->next, list2);
        return list1;
    }
    #endif
};


void test21() {
    #ifdef INFO
    cout << "\033[35m=========== INFO ===========\033[0m" << endl;
    cout << "FILE: " << __FILE__ << endl;
    cout << "APPROACH " << APR << ": " << apr_idea << endl;
    cout << "  + Time Complexity: O(" << time_cmplx << ")" << endl;
    cout << "  + Space Complexity: O(" << space_cmplx << ")" << endl;
    #endif

    cout << "\033[35m========== TESTCASE ========\033[0m\n";
    struct Case {
        // Inputs
        vector<int> l1_arr;
        vector<int> l2_arr;
        // Expect
        vector<int> exp;
    };

    vector<Case> cases = {
        // {{}}
        {{1,2,4}, {1,3,4}, {1,1,2,3,4,4}},
        {{}, {}, {}},
        {{}, {0}, {0}},
        {{2}, {1}, {1,2}},
        {{1,3,5}, {2,4,6}, {1,2,3,4,5,6}},
        {{1,2,3}, {}, {1,2,3}},
        {{}, {1,2,3}, {1,2,3}},
        {{5,6,7}, {1,2,3}, {1,2,3,5,6,7}},
        {{1,1,1}, {1,1,1}, {1,1,1,1,1,1}}
    };

    for (int i = 0; i < (int)cases.size(); ++i) {
        Solution sol = Solution();
        Case c = cases[i];
        ListNode* l1 = ListNode::createList(c.l1_arr);
        ListNode* l2 = ListNode::createList(c.l2_arr);
        ListNode* exp = ListNode::createList(c.exp);
        // Inputs
        auto in1 = l1;
        auto in2 = l2;

        // Result
        auto res = sol.mergeTwoLists(l1, l2);

        if(assertTest(res, exp, i) == false) {
            cout << "   Input    : " << in1 << " " << in2 << endl;
        }
    }
}
